docs/userguide/src/portingguide/concerns/weakref.md

```diff
@@ -0,0 +1,477 @@
+    optionally demand the object to be retained.  That object *and all descendants* will be kept
+    alive during this GC.
+    +   Do this with the `tracer_context` argument of `process_weak_refs`.
+-   **Request another invocation**: The VM binding can request `Scanning::process_weak_refs` to be
+    called again after computing the transitive closure that includes *retained objects and their
+    descendants*.  This helps handling multiple levels of weak reference strength.
+    +   Do this by returning `true` from `process_weak_refs`.
+
+The `Scanning::process_weak_refs` function also gives the VM binding a chance to perform other
+operations, including (but not limited to)
+
+-   **Do clean-up operations**: The VM binding can perform clean-up operations, or queue them to be
+    executed after GC.
+-   **update fields** that contain weak references.
+    -   **Forward the field**: It can write the forwarded address of the referent if moved by a
+        copying GC.
+    -   **Clear the field**: It can clear the field if the referent has not been reached and the
+        binding decides it is unreachable.
+
+Using those primitive operations, the VM binding can support different flavors of finalizers and/or
+weak references.  We will discuss common use cases in the following sections.
+
+## Supporting finalizers
+
+Different VMs define "finalizer" differently, but they all involve performing operations when an
+object is dead.  The general way to handle finalizer is visiting all **finalizable objects** (i.e.
+objects that have associated finalization operations), check if they are unreachable and, if
+unreachable, do something about them.
+
+### Identifying finalizable objects
+
+Some VMs determine whether an object is finalizable by its type.  In Java, for example, an object is
+finalizable if its `finalize()` method is overridden.  The VM binding can maintain a list of
+finalizable objects, and register instances of such types into that list when they are constructed.
+
+Some VMs can dynamically attach finalizing operations to individual objects after objects are
+created.  The VM binding can maintain a list of objects with attached finalizers, or maintain a
+(weak) hash map that maps finalizable objects to its associated finalizers.
+
+### When to run finalizers?
+
+Depending on the finalizer semantics in different VMs, finalizers can be executed during GC or
+during mutator time after GC.
+
+The VM binding can run finalizers immediately in `Scanning::process_weak_refs` when finding a
+finalizable object unreachable.  Beware that executing finalizers can be time-consuming.  The VM
+binding can creating work packets and let each work packet process a part of all finalizable
+objects.  In this way, multiple GC workers can process finalizable objects in parallel.  The
+`Scanning::process_weak_refs` function is executed in the `VMRefClosure` stage, so the created work
```

Comment:
Should we be using the term "stage" for work buckets?

Reply:
That's our current terminology.  We used to call it "phase" in JikesRVM.  Currently the type name is `WorkBucketStage` defined in `src/scheduler/work_bucket.rs`.